### ISSUE TITLE
IK with ML

### ISSUE THREAD

**Issue by Monocubes-afk** [StalyTV/Quadruped-Dog-ino#7]
Create ML model to predict angles from given x y z coordinates and save well perfoming model and transformations.

`import numpy as np
import pandas as pd
import joblib
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import StandardScaler

# === Load models and scalers ===
x_scaler = joblib.load('/content/drive/MyDrive/IKML/Model/x_scaler.pkl')
y_scaler = joblib.load('/content/drive/MyDrive/IKML/Model/y_scaler.pkl')

model_theta1 = joblib.load('/content/drive/MyDrive/IKML/Model/best_lgbm_theta1.pkl')
model_theta2 = joblib.load('/content/drive/MyDrive/IKML/Model/best_lgbm_theta2.pkl')
model_theta3 = joblib.load('/content/drive/MyDrive/IKML/Model/best_lgbm_theta3.pkl')

# === Forward Kinematics Function ===
def forward_kinematics(th1, th2, th3, L0=37.4, L1=80):
    t1 = np.radians(th1)
    t2 = np.radians(th2)
    t3 = np.radians(th3)

    x = -L1 * (np.sin(t2) * np.cos(t3) + np.sin(t3) * np.cos(t2)) - L1 * np.sin(t2)
    y = (L0 * np.cos(t1)
         - L1 * (np.sin(t1) * np.sin(t2) * np.sin(t3)
                 - np.sin(t1) * np.cos(t2) * np.cos(t3))
         + L1 * np.sin(t1) * np.cos(t2))
    z = (L0 * np.sin(t1)
         - L1 * (-np.sin(t2) * np.sin(t3) * np.cos(t1)
                 + np.cos(t1) * np.cos(t2) * np.cos(t3))
         - L1 * np.cos(t1) * np.cos(t2))
    return x, y, z

# === Inverse Kinematics Prediction Function ===
def predict_theta(x, y, z):
    input_scaled = x_scaler.transform([[x, y, z]])
    pred1 = model_theta1.predict(input_scaled)[0]
    pred2 = model_theta2.predict(input_scaled)[0]
    pred3 = model_theta3.predict(input_scaled)[0]
    pred_scaled = np.array([[pred1, pred2, pred3]])
    return y_scaler.inverse_transform(pred_scaled)[0]

# === Generate 200 Random Angle Sets ===
np.random.seed(42)
num_samples = 200
theta1_true = np.random.uniform(0, 40, num_samples)
theta2_true = np.random.uniform(-90, 0, num_samples)
theta3_true = np.random.uniform(0, 90, num_samples)

# === Run FK and then Predict ===
true_angles = []
pred_angles = []

for th1, th2, th3 in zip(theta1_true, theta2_true, theta3_true):
    x, y, z = forward_kinematics(th1, th2, th3)
    predicted = predict_theta(x, y, z)
    true_angles.append([th1, th2, th3])
    pred_angles.append(predicted)

true_angles = np.array(true_angles)
pred_angles = np.array(pred_angles)

# === Evaluation ===
mse = mean_squared_error(true_angles, pred_angles)
r2 = r2_score(true_angles, pred_angles)

print("\n🔍 Evaluation on 200 Random Test Samples")
print(f"Mean Squared Error (MSE): {mse:.4f}")
print(f"R² Score: {r2:.4f}")

# === Optional: Show comparison table ===
df_compare = pd.DataFrame({
    'θ1_true': true_angles[:, 0],
    'θ1_pred': pred_angles[:, 0],
    'θ2_true': true_angles[:, 1],
    'θ2_pred': pred_angles[:, 1],
    'θ3_true': true_angles[:, 2],
    'θ3_pred': pred_angles[:, 2],
})

print("\n📋 Sample Predictions:")
print(df_compare.head(10).round(2))
`

Code to test performance of the model or as a Template to create code to do IK and trajectories.
